### ISSUE TITLE
iOS Adomb crash -Multiple locks on web thread not allowed

### ISSUE THREAD

**Issue by zettyfactory** [kloder-games/godot-admob#53]
**OS target (Android/iOS):**
iOS

**Godot version:**
3.0.6

**Issue description:**
Everything works well in android.  
However, iOS has problems.
For example, an error occurs after Reward Video is terminated.

2018-11-13 04:30:20.350877+0900 hoonword[4062:184752] [avas] AVAudioSessionPortImpl.mm:56:ValidateRequiredFields: Unknown selected data source for Port Speaker (type: Speaker)
2018-11-13 04:30:20.374741+0900 hoonword[4062:185038] void _WebThreadLock(), 0x10c2ef040: Multiple locks on web thread not allowed! Please file a bug. Crashing now...
(lldb) 

I am struggling with this problem for a few days, but I am not an iOS developer, so it is difficult. I would be very grateful if you gave me the answer.

**Comment by zettyfactory** [comment]
I also asked the module developer, but I have not heard the answer yet. What bugs? Where should I fix it?

**Comment by Shin-NiL** [comment]
Which versions of Google Mobile Ads SDK, Xcode and iOS are you using?

Unfortunately I don't have any Apple hardware, so I'm afraid I can't help much here. I hope somebody from the community can help us to maintain this module for iOS.

**Comment by vinchi9** [comment]
I have encountered similar problem, on Android it works fine, but on iOS after terminating Interstitial Ad the application exits. Unfortunately I am not an iOS developer too and I don't own Mac or any iOS device, so debugging iOS apps is pretty hard for me.

Version Info:
Godot 3.0.6
XCode 10.1 (10B61)
Admob SDK 7.36
iPhone 6 (iOS 11.4)

**Comment by Shin-NiL** [comment]
Yeah, I need help from any iOS developer from the community to figure out a solution :(

**Comment by rolandoislas** [comment]
UIKit view modifications should be made only on the main thread. NSThread's [isMainThread](https://developer.apple.com/documentation/foundation/nsthread/1412704-ismainthread?language=objc) can be used to determine if the current thread is the main. If the current thread is not the main, the modifications can be delegated to it via [Dispatch](https://developer.apple.com/documentation/dispatch?language=objc) `dispatch_async` or `dispatch_sync_f`, using `dispatch_get_main_queue` as the queue.

**Comment by gvdb** [comment]
I was getting this problem too when returning from a RewardAd. I used the code from the examples provided in this repo and the crash seemed to be occurring on the label text update i.e. `get_node("CanvasLayer/LblRewarded").set_text("Reward: " + currency + ", " + str(amount))`

> UIKit view modifications should be made only on the main thread. NSThread's [isMainThread](https://developer.apple.com/documentation/foundation/nsthread/1412704-ismainthread?language=objc) can be used to determine if the current thread is the main. If the current thread is not the main, the modifications can be delegated to it via [Dispatch](https://developer.apple.com/documentation/dispatch?language=objc) `dispatch_async` or `dispatch_sync_f`, using `dispatch_get_main_queue` as the queue.

I wasn't sure how I do any of that within Godot, but the closest thing I could find that seemed in the right ballpark was the `[call_deferred](https://docs.godotengine.org/en/3.0/classes/class_object.html?highlight=call_deferred#class-object-call-deferred)` method. I wrapped the label text update code (above) in a `call_deferred` method and it seems to have fixed it, though I'm not entirely sure why.

Is there a better way to fix this within Godot? I don't understand how we would modify a Godot app using Apple's libraries (unless there are "hooks" to them within the Godot API?).

**Comment by Shin-NiL** [comment]
>I wrapped the label text update code (above) in a call_deferred method and it seems to have fixed it, though I'm not entirely sure why.

Thank you @gvdb, this is interesting information. Can you guys (@zettyfactory , @vinchi9 , @rolandoislas) try this workaround? If it really works, we can try fix the code.

**Comment by gvdb** [comment]
> > I wrapped the label text update code (above) in a call_deferred method and it seems to have fixed it, though I'm not entirely sure why.
> 
> Thank you @gvdb, this is interesting information. Can you guys (@zettyfactory , @vinchi9 , @rolandoislas) try this workaround? If it really works, we can try fix the code.

I tested it a little more this morning, and it still appears to be happening. Seems to be an intermittent issue. By luck I must've had a few instances where it didn't crash last night.

Will investigate more.

**Comment by gvdb** [comment]
I can't find any way to fix this within the Godot application.

Do the changes suggested by @rolandoislas need to be implemented in the code of this repo itself?

I've attempted to make changes to `AdmobRewarded.mm` using [these](https://stackoverflow.com/questions/39307800/how-to-check-current-thread-in-swift-3) [links](https://stackoverflow.com/questions/17490286/does-dispatch-asyncdispatch-get-main-queue-wait-until-done) as guidance, but my knowledge of the Apple code base is even worse than my knowledge of Godot. When I try to recompile the Godot engine source, I get errors I'm not sure how to fix.

Will keep trying and update if I make any progress.

**Comment by rolandoislas** [comment]
I think the loading of the ad needs to be done on the main thread. For example, [this line](https://github.com/kloder-games/godot-admob/blob/6c3fa503decf55aa4bc2e6b2a2881439cffb581a/admob/ios/src/AdmobInterstitial.mm#L42). I do not have hardware on hand to test at the moment, but I will in a week.

**Comment by gvdb** [comment]
> I think the loading of the ad needs to be done on the main thread. For example, [this line](https://github.com/kloder-games/godot-admob/blob/6c3fa503decf55aa4bc2e6b2a2881439cffb581a/admob/ios/src/AdmobInterstitial.mm#L42). I do not have hardware on hand to test at the moment, but I will in a week.

If you can give an example of the code that I would need to add to that line to get it to run on the main thread, I'm happy to test that on my iPhone now.

**Comment by rolandoislas** [comment]
Something like this should work.

```obj-c
#include <Foundation/Foundation.h>
#include <dispatch/dispatch.h>
     
if (![NSThread isMainThread]) {
    dispatch_async(dispatch_get_main_queue(), ^{
        [interstitial loadRequest:request];
    });
}
else {
    [interstitial loadRequest:request];
}
```

**Comment by gvdb** [comment]
Ok so I changed the loading code for banner ads and reward ads like so:

`    if (![NSThread isMainThread]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [bannerView loadRequest:request];
        });
    }
    else {
        [bannerView loadRequest:request];
    }`

`    if(!isReal) {
        if (![NSThread isMainThread]) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [[GADRewardBasedVideoAd sharedInstance] loadRequest:[GADRequest request]
                                                       withAdUnitID:@"ca-app-pub-3940256099942544/1712485313"];
            });
        }
        else {
            [[GADRewardBasedVideoAd sharedInstance] loadRequest:[GADRequest request]
                                                   withAdUnitID:@"ca-app-pub-3940256099942544/1712485313"];
        }
    }
    else {
        if (![NSThread isMainThread]) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [[GADRewardBasedVideoAd sharedInstance] loadRequest:[GADRequest request] withAdUnitID:rewardedId];
            });
        }
        else {
            [[GADRewardBasedVideoAd sharedInstance] loadRequest:[GADRequest request] withAdUnitID:rewardedId];
        }
    }`

Then I followed the steps for installing the admob module (i.e. recompiling Godot source etc.). The game I've built has a banner ad and a reward ad in it and the web lock crash still occurs when exiting the reward ad.

**Comment by Shin-NiL** [comment]
@gvdb Well it's basically what I did. Can you tell me if you don't use banner the app still crash?
Maybe we need to instance the adds on the main thread, not only the loading.

**Comment by gvdb** [comment]
Yep, it all works fine when I remove the banner ad. I tested exiting the video early, and allowing it play all the way through to get the reward. Both were fine.

**Comment by Shin-NiL** [comment]
@gvdb could you try my new commit https://github.com/Shin-NiL/godot-admob/tree/issue_53? I'm trying to putting everything on the main thread (similar to what happens in the Android code). The code is very ugly, but first we need to know if it works.

**Comment by gvdb** [comment]
Got this when trying to compile:

```
[ 11%] Compiling ==> modules/admob/ios/src/godotAdmob.mm
modules/admob/ios/src/godotAdmob.mm:175:13: error: return type 'uintptr_t' (aka 'unsigned long') must match previous return type 'int' when block literal has unspecified explicit return type [2]
             return (uintptr_t)[banner getBannerWidth];
             ^
modules/admob/ios/src/godotAdmob.mm:170:9:{170:9-170:23}: error: no matching function for call to 'dispatch_async' [2]
         dispatch_async(dispatch_get_main_queue(), ^{
         ^~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/dispatch/queue.h:219:1: note: candidate function not viable: no known conversion from 'int (^)()' to 'dispatch_block_t _Nonnull' (aka 'void (^)()') for 2nd argument [2]
 dispatch_async(dispatch_queue_t queue, dispatch_block_t block);
 ^
modules/admob/ios/src/godotAdmob.mm:195:13: error: return type 'uintptr_t' (aka 'unsigned long') must match previous return type 'int' when block literal has unspecified explicit return type [2]
             return (uintptr_t)[banner getBannerHeight];
             ^
modules/admob/ios/src/godotAdmob.mm:190:9:{190:9-190:23}: error: no matching function for call to 'dispatch_async' [2]
         dispatch_async(dispatch_get_main_queue(), ^{
         ^~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/dispatch/queue.h:219:1: note: candidate function not viable: no known conversion from 'int (^)()' to 'dispatch_block_t _Nonnull' (aka 'void (^)()') for 2nd argument [2]
 dispatch_async(dispatch_queue_t queue, dispatch_block_t block);
 ^
4 errors generated.
scons: *** [modules/admob/ios/src/godotAdmob.iphone.opt.debug.arm64.o] Error 1
scons: building terminated because of errors.
```

**Comment by Shin-NiL** [comment]
getBannerWidth and height should not be in the main thread, reverted. Could you try again, please?

**Comment by gvdb** [comment]
Ok, it compiles now, but the crash still occurs.

**Comment by cagdasc** [comment]
Are there any changes?

**Comment by Shin-NiL** [comment]
@cagdasc nope, seems like there is not so much iOS devs using Godot :(

**Comment by wombatwingdings** [comment]
This problem isn't just related to reward video. I get exactly the same crash (Multiple locks on web thread not allowed) on **Interstitial** when used with banner ads even if I switch banner ad off before displaying.

**Comment by cagdasc** [comment]
@Shin-NiL I try it with different threads but result is same. Maybe we need to try this scenario in native iOS app.

**Comment by Shin-NiL** [comment]
@cagdasc unfortunately nothing we've tried has worked so far ;(

**Comment by cagdasc** [comment]
@Shin-NiL I taken this crash on production environment for my game and I had to remove interstitial ads. There is no difference between simulator or real devices.

**Comment by Shin-NiL** [comment]
Please, can someone test [this branch](https://github.com/Shin-NiL/godot-admob/tree/issue_53-2)? I've changed how the module gets the rootController.

**Comment by wombatwingdings** [comment]
@Shin-NiL I have tested your branch. Rebuilt the export templates, but unfortunately, no difference:

![image](https://user-images.githubusercontent.com/36036908/52241338-443f9300-28cb-11e9-9f0e-9b8ee634d285.png)

Sorry, but thanks for trying!

**Comment by wombatwingdings** [comment]
@Shin-NiL , glad to be able to help in some small way.
I'm afraid new version crashes in the same place too.

**Comment by Shin-NiL** [comment]
Thanks again @wombatwingdings!
It's really weird, because a friend of mine have told me it worked fine but on simulator and using godot 2.1.5. Anyways, I'll merge these changes to master, as it seems to not broke anything.

**Comment by kynora** [comment]
I solve this problem in GDScript by call resize to remove banner from view before show Reward or Interstitial. And after banner ad loaded check to show banner if either Interstitial or Reward not showing.

func _on_admob_ad_loaded():
   if interstitialShowing || rewardShowing:
      yield(get_tree().create_timer(5.0), "timeout")
      _on_admob_ad_loaded()
   else:
      showBanner()

**Comment by wombatwingdings** [comment]
@kynora, what class did you call resize() on to remove the banner?

**Comment by kynora** [comment]
@wombatwingdings, resize() call from GodotAdMob class. In code when banner resize it remove from view and load request again.
Here is my GDScript to show rewarded video:
```
func showRewardVideo():
   if admob:
      rewardShowing = true
      bannerShowing = false
      admob.resize()
      admob.showRewardedVideo()
```

**Comment by kynora** [comment]
@wombatwingdings , full GDScript: 
```

extends Node
var isReal = true

var admob = null
var isTop = true


# bug with closing interstitial and reward crash with banner
var rewardLoaded = false
var intLoaded = false

var bannerShowing = false
var rewardShowing = false
var intShowing = false

var bannerId
var interstitialId
var rewardedId

func _ready():
   initAdMob()

func initAdMob():

   if Engine.has_singleton("GodotAdMob"):
      admob = Engine.get_singleton("GodotAdMob")
      admob.init(isReal, get_instance_id())
   loadBanner()
   loadInterstitial()
   loadRewardedVideo()
   
   get_tree().connect("screen_resized", self, "onResize")

func loadBanner():
   var bannerShowing = false
   if admob:
      admob.loadBanner(bannerId, isTop)
      

func loadInterstitial():
   intShowing = false
   intLoaded = false
   if admob:
      admob.loadInterstitial(interstitialId)
      
func loadRewardedVideo():
   rewardShowing = false
   rewardLoaded = false
   if admob:
      admob.loadRewardedVideo(rewardedId)

func showBanner():
   if bannerShowing:
      return
   if admob:
         bannerShowing = true
         admob.showBanner()
      
func showInterstitial():
   if admob:
      if intLoaded:
         intShowing = true
         bannerShowing = false
         admob.resize()
         admob.showInterstitial()

   
func showRewardVideo():
   if admob:
      rewardShowing = true
      bannerShowing = false
      admob.resize()
      admob.showRewardedVideo()


func _on_admob_network_error():
   print("Network Error")
   yield(get_tree().create_timer(20.0), "timeout")
   loadBanner()
   

func _on_admob_ad_loaded():
   print("Ad loaded success")
   if intShowing || rewardShowing:
      yield(get_tree().create_timer(5.0), "timeout")
      _on_admob_ad_loaded()
   else:
      showBanner()

func _on_interstitial_not_loaded():
   print("Error: Interstitial not loaded")
   yield(get_tree().create_timer(20.0), "timeout")
   loadInterstitial()

func _on_interstitial_loaded():
   print("Interstitial loaded")
   intLoaded = true
   

func _on_interstitial_close():
   print("Interstitial closed")
   loadInterstitial()


func _on_rewarded_video_ad_failed_to_load():
   yield(get_tree().create_timer(20.0), "timeout")
   loadRewardedVideo()

func _on_rewarded_video_ad_loaded():
   print("Rewarded loaded success")
   rewardLoaded = true
   
func _on_rewarded_video_ad_closed():
   print("Rewarded closed")
   loadRewardedVideo()
   

   
func _on_rewarded(currency, amount):
   print("Reward: " + currency + ", " + str(amount))

```

**Comment by Shin-NiL** [comment]
@kynora for the first time it makes sense. So the conflict seems to be with the banner on screen. Did you try only hideBanner() instead of resize()?

**Comment by kynora** [comment]
I not yet try hideBanner(). Because hideBanner is only hidden image of ads but view still remaining. Other solution is implement all ads in ViewController of engine, but it is not following with engine workflow.

**Comment by Shin-NiL** [comment]
Could you guys please test the latest commit for ou friend @yamshing? It should finally fix this issue.

**Comment by Dridia1** [comment]
Doesn't seem like the issue has been solved. I've recreated the same issue with what @yamshing changed.

**Comment by yamshing** [comment]
@Dridia1 Could you test with delay bigger than 0 at
https://github.com/kloder-games/godot-admob/blob/master/admob/ios/src/AdmobRewarded.mm#L83
? If it does nothing tell me more about your environnement ( ios version etc... simulator or real device ?)

**Comment by Dridia1** [comment]
@yamshing I were using the interstitial ad, so I changed the following line instead:
https://github.com/kloder-games/godot-admob/blob/f1dcaa3895973b5ca2d014d331950b9cade9335c/admob/ios/src/AdmobInterstitial.mm#L89

This works perfect, and I'm running iOS 12.4 on an iPhone X

**Comment by yamshing** [comment]
@Dridia1 thanks for your test. I will change the delay and commit the change

**Comment by canvasbushi** [comment]
you should also wrap your bannerEnable method with a

dispatch_async(dispatch_get_main_queue(), ^{
        // enable banner
    });

so the ui gets updated on the main thread.

**Comment by yamshing** [comment]
@canvasbushi thank you for your suggestion I will test this

**Comment by gustavottc** [comment]
Some one can test?

I tested on #124 , and its working fine!

Ps: version 3.1.2, version 3.2.1 i doesnt test yet

**Comment by gustavottc** [comment]
This issue is fixed on 3.1.2, if someone is still facing this issue, please comment here

**Comment by pchasco** [comment]
On 2.1 I continue to get this error. In fact, I have stripped everything from the module except the bare minimum of code required by Google to implement a rewarded ad and still receive the error. The issue must be in the engine itself. I have also tried with both Google Ads Frameworks versions 7.41 and 7.60. Same result.

**Comment by pchasco** [comment]
Godot 3 should now have this issue fixed, as a patch has been merged. The fix has also been picked into the 2.1.7 branch, though at this time it hasn't been officially released. I manually applied the fixes to my own Godot fork.

**Comment by gustavottc** [comment]
Hey @pchasco , yes, on Godot 3 this issue is fixed, i dont know if 2.1.X still happens.

2.1.X version is still in working by Godot contributors?

**Comment by pchasco** [comment]
I can’t say for sure, but the patch has been cherry picked for 2.1.7, so
it’s a possibility.
